rslib/src/cloze.rs

```diff
@@ -31,99 +19,262 @@ lazy_static! {
     .unwrap();
 }
 
-mod cloze_caps {
-    // cloze ordinal
-    pub const ORD: usize = 1;
-    // the occluded text
-    pub const TEXT: usize = 2;
-    // optional hint
-    pub const HINT: usize = 3;
-}
-
 mod mathjax_caps {
     pub const OPENING_TAG: usize = 1;
     pub const INNER_TEXT: usize = 2;
     pub const CLOSING_TAG: usize = 3;
 }
 
-pub fn reveal_cloze_text(text: &str, cloze_ord: u16, question: bool) -> Cow<str> {
-    let mut cloze_ord_was_in_text = false;
-
-    let output = CLOZE.replace_all(text, |caps: &Captures| {
-        let captured_ord = caps
-            .get(cloze_caps::ORD)
-            .unwrap()
-            .as_str()
-            .parse()
-            .unwrap_or(0);
-
-        let text = caps.get(cloze_caps::TEXT).unwrap().as_str().to_owned();
-        if captured_ord != cloze_ord {
-            // other cloze deletions are unchanged
-            return text;
-        } else {
-            cloze_ord_was_in_text = true;
+/// States for cloze parsing state machine
+#[derive(PartialEq, Copy, Clone)]
+enum State {
+    Root,
+    Text,
+    Open,
+    Open2,
+    COpen,
+    Ord,
+    TextOpen1,
+    TextOpen2,
+    TextClose1,
+    HintOpen1,
+    HintOpen2,
+    Hint,
+    HintClose1,
+    Close,
+    Abandon,
+}
+/// Struct for storing data for one cloze
+struct Cloze {
+    buf: String,
+    ord_str: String,
+    text: String,
+    hint: String,
+}
+
+impl Cloze {
+    /// Create empty Cloze struct
+    fn new() -> Self {
+        Self {
+            buf: String::new(),
+            ord_str: String::new(),
+            text: String::new(),
+            hint: String::new(),
+        }
+    }
+    /// Create Cloze struct with buf set to c
+    fn new_from(c: char) -> Self {
+        Self {
+            buf: String::from(c),
+            ord_str: String::new(),
+            text: String::new(),
+            hint: String::new(),
         }
+    }
+}
+
+/// Process char for next state
+fn process_char(state: State, c: char) -> State {
+    use self::State::*;
+    match (state, c) {
+        (Root, '{') => Open,
+        (Open, '{') => Open2,
```

Comment:
Processing a character a time feels a bit complicated. I wonder whether this could be simplified by tokenizing the string with `nom` instead?  We already use it for dealing parsing card templates.

https://docs.rs/nom/latest/nom/

Reply:
Possibly. I don't feel the state machine is particularly complicated (although probably a bit more verbose compared to a library). I am unfamiliar with Rust and its libraries so it was simpler to write a state machine. Are you adamant on using `nom`?

Reply:
> I am unfamiliar with Rust

You've done a good job - I can't really tell.

I'd appreciate it if you'd give nom a try - I've done the bulk of the parsing work for you, so it should mostly be a case of dealing with the tokens now.

```rust
use std::error::Error;

use nom::branch::alt;
use nom::bytes::complete::{take_until, take_while};
use nom::multi::many0;
use nom::{bytes::complete::tag, combinator::map, IResult};

#[derive(Debug)]
enum Token<'a> {
    OpenCloze(u16),
    Text(&'a str),
    Hint(&'a str),
    CloseCloze,
}

fn open_cloze(text: &str) -> IResult<&str, Token> {
    // opening brackets and 'c'
    let (text, _opening_brackets_and_c) = tag("{{c")(text)?;
    // following number
    let (text, digits) = take_while(|c: char| c.is_ascii_digit())(text)?;
    let digits: u16 = match digits.parse() {
        Ok(digits) => digits,
        Err(_) => {
            // not a valid number; fail to recognize
            return Err(nom::Err::Error(nom::error::make_error(
                text,
                nom::error::ErrorKind::Digit,
            )));
        }
    };
    // ::
    let (text, _colons) = tag("::")(text)?;
    Ok((text, Token::OpenCloze(digits)))
}

fn close_cloze(text: &str) -> IResult<&str, Token> {
    map(tag("}}"), |_| Token::CloseCloze)(text)
}

fn hint(text: &str) -> IResult<&str, Token> {
    let (text, _separating_colons) = tag("::")(text)?;
    let (text, hint) = take_until("}}")(text)?;
    Ok((text, Token::Hint(hint)))
}

/// Match a run of text until an open/close or hint is encountered.
/// This will stop on a hint marker even outside a cloze, so the processing
/// should handle hint tokens outside of an active cloze.
fn normal_text(text: &str) -> IResult<&str, Token> {
    if text.is_empty() {
        return Err(nom::Err::Error(nom::error::make_error(
            text,
            nom::error::ErrorKind::Eof,
        )));
    }
    let mut index = 0;
    let mut other_token = alt((open_cloze, close_cloze, hint));
    while other_token(&text[index..]).is_err() && index + 1 < text.len() {
        index += 1;
    }
    Ok((&text[index..], Token::Text(&text[0..index])))
}

// todo: error handling
fn tokenize(text: &str) -> Vec<Token> {
    let (remaining, tokens) =
        many0(alt((open_cloze, hint, close_cloze, normal_text)))(text).unwrap();
    assert!(remaining.is_empty());
    tokens
}

fn main() -> Result<(), Box<dyn Error>> {
    dbg!(tokenize("foo {{c1::bar {{c2::baz}}}}"));
    dbg!(tokenize("foo {{c1::bar {{c2::baz}}::qux}}"));
    Ok(())
}
```

Example output:

```
[src/main.rs:74] tokenize("foo {{c1::bar {{c2::baz}}::qux}}") = [
    Text(
        "foo ",
    ),
    OpenCloze(
        1,
    ),
    Text(
        "bar ",
    ),
    OpenCloze(
        2,
    ),
    Text(
        "baz",
    ),
    CloseCloze,
    Hint(
        "qux",
    ),
    CloseCloze,
]
```

Reply:
(that should give two advantages - it refers to the original strings of constructing a bunch of new ones a character a time, and it allows the parsing code to deal with only 4 token kinds instead of having to ignore a bunch of other ones like HintClose1)

Reply:
Found a bug; this line should have the + 1 removed from it:

```
    while other_token(&text[index..]).is_err() && index < text.len() {
```

Reply:
Are you ok with me changing the design of `normal_text` and the use of `nom::alt` to keep track of state? I don't immediately see a clean way of resolving the tokens for input like `foo }} :: {{c1::bar}} baz` (which is admittedly weird but who knows what people want to put in their clozes) as it results in something like

```
Text(
    "foo ",
),
CloseCloze,
Text(
" ",
),
Hint(
    " {{c1::bar",
),
CloseCloze,
Text(
    " baz",
)
```

Reply:
Hmm, good point about hint being problematic there. I was hoping to keep the tokenizing as state-free as possible to keep it simple, but it shouldn't be at the expense of correctness. Maybe one alternative to keeping state in normal_text would be to remove the hint token from the tokenizer, and only focus on opening/closing/text? Then once the state machine has identified a well-formed cloze deletion, it could extract a hint from it if one exists.

Reply:
Am I correct in that your end goal is simple/non-verbose code from a maintenance perspective? From a performance perspective one could make the state machine use slices/indexes rather than pushing to different `String`s which would make it zero copy. I doubt that any multiple pass solution will be faster than a single pass over the input string. Granted, I don't think that this piece of code will be a bottle neck in the overall scheme of things.

Reply:
Yes, ease of maintenance is my primary concern - if we have to come back to this code months or years down the line and make changes to it, I'm more confident it will be easy if the tokenizing and cloze construction is separated and can be reasoned about separately. Performance is less of a concern, though constructing strings a character at a time isn't ideal - when the user does something like 'check media', every note's clozes need to be rendered, so an inefficient solution could end up with noticeably slower results.

Reply:
Ok, so I will look at making a two stage process, first tokenize clozes and then parse out any hints from found clozes? 

For reference, as I read it `normal_text()` results in `text.len() + 1` runs of `alt((open_cloze, close_cloze))`, each of which mean at least 1-2 iterations over the "remaining string" so there are a significant number of iterations. As earlier mentioned I think the state machine modified to use `&str`/indices will be significantly faster if speed is a concern as it is a single iteration.

Reply:
Yes please. Fetching the hint is probably just a case of s.rsplitn(2, "::"), so you could probably defer doing it until you need to render/extract the text. Re performance, the open/close nom filters should return after checking the first character if it's not a { or }, so I suspect the overhead over a bespoke state machine would mainly be the function invocation/Result return (which may end up getting inlined by the compiler anyway). If it proves to be an issue, it would be fairly simple to try those two filters only after manually checking the first character.

Reply:
Ok, so I've looked and I have a hard time coming up with a clean solution without any form of state, the following isn't proofed yet but an idea of what I am thinking:

``` Rust
use std::{error::Error, borrow::Cow};

use nom::branch::alt;
use nom::bytes::complete::{take_until, take_while};
use nom::multi::many0;
use nom::{bytes::complete::tag, combinator::map, IResult};

#[derive(Debug, Copy, Clone)]
enum Token<'a> {
    OpenCloze(u16),
    Text(&'a str),
    OpenHint,
    CloseCloze,
}

fn open_cloze(text: &str) -> IResult<&str, Token> {
    // opening brackets and 'c'
    let (text, _opening_brackets_and_c) = tag("{{c")(text)?;
    // following number
    let (text, digits) = take_while(|c: char| c.is_ascii_digit())(text)?;
    let digits: u16 = match digits.parse() {
        Ok(digits) => digits,
        Err(_) => {
            // not a valid number; fail to recognize
            return Err(nom::Err::Error(nom::error::make_error(
                text,
                nom::error::ErrorKind::Digit,
            )));
        }
    };
    // ::
    let (text, _colons) = tag("::")(text)?;
    Ok((text, Token::OpenCloze(digits)))
}

fn open_hint(text: &str) -> IResult<&str, Token> {
    map(tag("::"), |_| Token::OpenHint)(text)
}

fn content(text: &str) -> IResult<&str, Token> {
    alt((open_hint, close_cloze, open_cloze))(text)
}

fn close_cloze(text: &str) -> IResult<&str, Token> {
    map(tag("}}"), |_| Token::CloseCloze)(text)
}

fn open_or_close_cloze(text: &str) -> IResult<&str, Token> {
    alt((close_cloze, open_cloze))(text)
}


fn tokenize(mut text: &str) -> Result<Vec<Token>, u32> {
    let mut tokens = Vec::new();
    let mut token = Token::CloseCloze;
    let mut depth = 0;

    while !text.is_empty() {
        let pattern = match (token, depth > 1) {
            (Token::CloseCloze, false) => {
                depth = 0;
                open_cloze
            },
            (Token::CloseCloze, true) => {
                depth -= 1;
                open_or_close_cloze
            },
            (Token::OpenCloze(_), _) => {
                depth += 1;
                content
            }
            (Token::OpenHint, _) => close_cloze,
            (_, _) => open_or_close_cloze
        };

        let mut i = 0;
        let mut done = false;
        while !done && i < text.len() {
            let found = pattern(&text[i..]);
            if found.is_ok() {
                if i > 0 { tokens.push(Token::Text(&text[..i])) };
                (text, token) = found.unwrap();
                tokens.push(token);
                done = true;
            } else {
                i += 1;
            }
        }
        if i == text.len() {
            tokens.push(Token::Text(&text[..i]));
            text = &text[i..];
        }
    }
    return Ok(tokens);
}

fn main() -> Result<(), Box<dyn Error>> {
    dbg!(tokenize(r#"foo {{c1::bar::q{{c1::ux}} baz"#).unwrap());
    Ok(())
}
```

Although there is state in the above in the form of "last token" I hope that this is sufficient.

Reply:
The approach I had in mind keeps the tokenizing process stateless, and keeps the logic in the subsequent parsing code, eg:

```rust
use std::error::Error;

use nom::branch::alt;
use nom::bytes::complete::{take_until, take_while};
use nom::multi::many0;
use nom::{bytes::complete::tag, combinator::map, IResult};

#[derive(Debug)]
enum Token<'a> {
    OpenCloze(u16),
    Text(&'a str),
    CloseCloze,
}

fn open_cloze(text: &str) -> IResult<&str, Token> {
    // opening brackets and 'c'
    let (text, _opening_brackets_and_c) = tag("{{c")(text)?;
    // following number
    let (text, digits) = take_while(|c: char| c.is_ascii_digit())(text)?;
    let digits: u16 = match digits.parse() {
        Ok(digits) => digits,
        Err(_) => {
            // not a valid number; fail to recognize
            return Err(nom::Err::Error(nom::error::make_error(
                text,
                nom::error::ErrorKind::Digit,
            )));
        }
    };
    // ::
    let (text, _colons) = tag("::")(text)?;
    Ok((text, Token::OpenCloze(digits)))
}

fn close_cloze(text: &str) -> IResult<&str, Token> {
    map(tag("}}"), |_| Token::CloseCloze)(text)
}

/// Match a run of text until an open/close marker is encountered.
fn normal_text(text: &str) -> IResult<&str, Token> {
    if text.is_empty() {
        return Err(nom::Err::Error(nom::error::make_error(
            text,
            nom::error::ErrorKind::Eof,
        )));
    }
    let mut index = 0;
    let mut other_token = alt((open_cloze, close_cloze));
    while other_token(&text[index..]).is_err() && index < text.len() {
        index += 1;
    }
    Ok((&text[index..], Token::Text(&text[0..index])))
}

// todo: error handling
fn tokenize(text: &str) -> Vec<Token> {
    let (remaining, tokens) = many0(alt((open_cloze, close_cloze, normal_text)))(text).unwrap();
    assert!(remaining.is_empty());
    tokens
}

#[derive(Debug)]
enum ParseOutput<'a> {
    Text(&'a str),
    Cloze(ExtractedCloze<'a>),
}

#[derive(Debug)]
struct ExtractedCloze<'a> {
    number: u16,
    text_fragments: Vec<&'a str>,
}

fn parse_tokens<'a>(tokens: &'a [Token]) -> Vec<ParseOutput<'a>> {
    let mut open_clozes: Vec<ExtractedCloze> = vec![];
    // nested clozes are placed here until the stack is empty, so they are gathered in the correct order
    let mut nested_clozes: Vec<ExtractedCloze> = vec![];
    let mut output = vec![];
    for token in tokens {
        match token {
            Token::OpenCloze(number) => open_clozes.push(ExtractedCloze {
                number: *number,
                text_fragments: vec![],
            }),
            Token::Text(text) => {
                if let Some(cloze) = open_clozes.last_mut() {
                    cloze.text_fragments.push(text);
                } else {
                    output.push(ParseOutput::Text(text));
                }
            }
            Token::CloseCloze => {
                let nested = open_clozes.len() > 1;
                if let Some(cloze) = open_clozes.pop() {
                    if nested {
                        nested_clozes.push(cloze);
                    } else {
                        output.push(ParseOutput::Cloze(cloze));
                        loop {
                            match nested_clozes.pop() {
                                Some(cloze) => output.push(ParseOutput::Cloze(cloze)),
                                None => break,
                            }
                        }
                    }
                } else {
                    output.push(ParseOutput::Text("}}"))
                }
            }
        }
    }
    output
}

fn main() -> Result<(), Box<dyn Error>> {
    dbg!(parse_tokens(&tokenize("foo {{c1::bar {{c2::baz}}}}")));
    dbg!(parse_tokens(&tokenize(
        "foo {{{c1::bar {{c2::baz}}::qux}} hello ::"
    )));
    dbg!(parse_tokens(&tokenize("foo }} :: {{c1::bar}} baz")));

    Ok(())
}
```

What do you think?

Reply:
As far as I can tell this still parses `foo }} :: {{c1::bar::pok{{c3::pok}} baz` incorrectly for example (the `{{c3::` inside the hint should be parsed as "hint text" i think)?

Reply:
Your earlier example could crop up in normal use when users use `::` outside of a cloze. I think this example is rather less likely - can't we just declare that opening cloze deletion markers is not valid inside the hint area?

Reply:
Admittedly my last example is a bit contrived, I was aiming for something that wouldn't interpret any existing notes incorrectly from a principal level.  As far as I can tell the basic "syntactical rules" for which  tokens should be parsed depends on the current state:

- "Root" text: "open cloze"
- "Open cloze": "open cloze", "open hint" and "close cloze"
- "Open hint": "close cloze"
- "Close cloze": "open cloze" and "close cloze" unless at "root level"

From my perspective a solution which takes this into account is cleaner (and more maintainable) but that is just a question of taste.

Reply:
Not sure about "more maintainable", as it's adding complexity for a somewhat unlikely case, but I don't have strong feelings about it if the change to parse_tokens() you're thinking of will still be easy to follow. Presumably it would just be a matter of looking for a '::' when pushing a text fragment onto the current cloze, and setting a flag that prevents opening new clozes until the current one is closed.

Reply:
I am not sure I understand but as mentioned I don't see a clean solution for `tokenize()`/`parse_tokens()` without some state during tokenizing (such as my earlier `tokenize()`example). If you are set on stateless tokenizing then I suggest going with your solution above and declaring "cloze open" invalid inside hints? I.e. I will look at a resolver for your `tokenizing()`/parse_tokens()`?

Reply:
I don't understand why tokenize would need changing - couldn't it be solved using something like this?

```rust
fn parse_tokens<'a>(tokens: &'a [Token]) -> Vec<ParseOutput<'a>> {
    let mut open_clozes: Vec<ExtractedCloze> = vec![];
    // nested clozes are placed here until the stack is empty, so they are gathered in the correct order
    let mut nested_clozes: Vec<ExtractedCloze> = vec![];
    let mut output = vec![];
    let mut current_cloze_has_hint = false;
    for token in tokens {
        match token {
            Token::OpenCloze(number) => {
                if current_cloze_has_hint {
                    // text_fragments would need to become a Cow for us to encode the original number
                    open_clozes
                        .last_mut()
                        .unwrap()
                        .text_fragments
                        .push("{{cx::")
                } else {
                    open_clozes.push(ExtractedCloze {
                        number: *number,
                        text_fragments: Vec::with_capacity(1), // common case
                    });
                }
            }
            Token::Text(text) => {
                if let Some(cloze) = open_clozes.last_mut() {
                    if text.contains("::") {
                        current_cloze_has_hint = true;
                    }
                    cloze.text_fragments.push(text);
                } else {
                    output.push(ParseOutput::Text(text));
                }
            }
            Token::CloseCloze => {
                let nested = open_clozes.len() > 1;
                if let Some(cloze) = open_clozes.pop() {
                    current_cloze_has_hint = false;
                    if nested {
                        nested_clozes.push(cloze);
                    } else {
                        output.push(ParseOutput::Cloze(cloze));
                        while let Some(cloze) = nested_clozes.pop() {
                            output.push(ParseOutput::Cloze(cloze))
                        }
                    }
                } else {
                    output.push(ParseOutput::Text("}}"))
                }
            }
        }
    }
    output
}
```

But that said, I'd lean towards just declaring this case as unsupported for now, as I suspect we don't actually need it.

Reply:
> I don't understand why tokenize would need changing - couldn't it be solved using something like this?

It can definitely be solved by something like your solution, I just meant I have a hard time coming up with a solution that I find "clean" (not several layers of parsing etc.), but that is just my personal taste. If you are ok with declaring `{{c1::` unsupported in the hint I will update the PR with your earlier suggestion (and the other issues) and add the "rendering" code.

Reply:
Yes please, let's declare it unsupported until we know we need it.

Reply:
Ok, I've updated the PR, I don't know if you wanted another layer of functions to take the result of `parse_tokens()` or to implement the rendering logic "inside" it. I went with the latter and moved `parse_tokens()` inside `reveal_cloze_text()` and `reveal_cloze_text_only()` as the logic the required steps are fewer in the latter. I also put "local local" functions and types inside their only user function (e.g. the different tokenizing functions) to make it clearer where they are and aren't used.

Sorry for all the back and forth on the "cloze open" inside a hint, it is now supported because it made sense in the way the logic turned out.

As I am no rust expert: someone who is good with rust memory management might want to look at string handling in `reveal_cloze_text()`, possibly there is some unnecessary string copying going on there (I had to create `String` from `&str` earlier than I would like). I can see some solutions to it but they involve growing the code so I don't know if a slight performance gain is worth a larger code base.